### PR TITLE
Test improvements

### PR DESCRIPTION
**Pull request by trentmwillis**
This improves some various tests. Most notably it adds tests to verify the order of execution for logging callbacks and module hooks (as raised in https://github.com/jquery/api.qunitjs.com/issues/136).

While writing these tests, 2 questions came up:

1. Should the "done" category of callbacks (`done`, `moduleDone`, and `testDone`) be FIFO or LIFO? I know we had a lot of discussion about a similar topic in #977.
2. Should `moduleDone` fire when starting a nested module? I don't believe it should, since the parent module is not technically finished yet.

### CODE REVIEW

**Comment by platinumazure** [comment]
For the second issue, I would say we should avoid firing moduleDone until it and all nested modules are truly finished. Otherwise that event conceivably fires twice (I know we've had an issue on that).

File: test/callbacks.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+			"module1 > before",
+			"module1 > beforeEach",
+			"module1 > test1",
+			"log1",
+			"log2",
+			"module1 > afterEach",
+			"testDone1",
+			"testDone2",
+			"moduleDone1", // This shouldn't happen, should it?
+			"moduleDone2", // This shouldn't happen, should it?
```
Comment (leobalter) [inline test/callbacks.js]: why not? these moduleDone are for the module that has started on [line 43](https://github.com/jquery/qunit/pull/1043/files#diff-e6fe7aebcef218e471b7b20028f2df99R43). This seems right

File: test/callbacks.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+			"log2",
+			"module2 > afterEach",
+			"module1 > afterEach",
+			"module2 > after",
+			"testDone1",
+			"testDone2",
+			"moduleDone1",
+			"moduleDone2",
+			"moduleStart1", // This shouldn't happen, should it?
+			"moduleStart2", // This shouldn't happen, should it?
```
Comment (leobalter) [inline test/callbacks.js]: same thing here, this seems right, related to the modules done at lines 86/87

File: test/callbacks.js (hunk 1 above)
Comment (trentmwillis) [inline test/callbacks.js]: The module that started on line 43 doesn't actually end until line 86/87.

**Comment by leobalter** [comment]
> Should moduleDone fire when starting a nested module? I don't believe it should, since the parent module is not technically finished yet.

IMHO, I believe it's ok to fire this event. That respective module is finished, not the parent module. The provided details should identify the module that is done.

Besides the comments above, this looks great. It's nice to have specific tests for the _execution order_ (which is also a good band name).

**Comment by trentmwillis** [comment]
@leobalter I suppose that depends on how you view modules. I consider nested modules to be part of their parent module. In other words, given:

```javascript
module('parent', function() {
  test('1', function() {});
  module('child', function() {
    test('1', function() {});
  });
  test('2', function() {});
});
```

I would expect the order of events to be:

```
parent -> moduleStart
child -> moduleStart
child -> moduleEnd
parent -> moduleEnd
```

Which reflects the structure of the code as well. What we have *currently* is:

```
parent -> moduleStart
parent -> moduleEnd
child -> moduleStart
child -> moduleEnd
parent -> moduleStart
parent -> moduleEnd
```

Which, in my opinion, should translate to the following code:

```javascript
module('parent', function() {
  test('1', function() {});
});
module('child', function() {
  test('1', function() {});
});
module('parent', function() {
  test('2', function() {});
});
```

Essentially, I don't think a module should be considered "done" until all of it's children (whether tests or other modules) are also done.

File: test/callbacks.js (hunk 1 above)
Comment (leobalter) [inline test/callbacks.js]: I finally got it. It's interesting to see how the module1's `after` hook isn't triggered, at least.

**Comment by leobalter** [comment]
```
parent -> moduleStart
parent -> moduleEnd
child -> moduleStart
child -> moduleEnd
parent -> moduleStart
parent -> moduleEnd
```

geez.

yes, this is not good, thanks for clarifying this part.

**Comment by leobalter** [comment]
It's easy to see the callbacks are following when a specific module is running, but the moduleDone is indeed expected right after it's respective `after` hook.

**Comment by platinumazure** [comment]
For what it's worth, as a QUnit user, I would much prefer @trentmwillis' interpretation.

**Comment by leobalter** [comment]
> For what it's worth, as a QUnit user, I would much prefer @trentmwillis' interpretation.

and that's where I agree and even consider the current way it works is a bug.

**Comment by trentmwillis** [comment]
Okay, I think we can merge this test and follow up on improving the `moduleDone` behavior in a separate PR/issue.

**Pull request closed without merging**